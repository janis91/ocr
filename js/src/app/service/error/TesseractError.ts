import { OcrError } from './OcrError';

export class TesseractError extends OcrError {
    constructor(message: string, public urlOrCanvas: string | HTMLCanvasElement, original?: Error | any) {
        super(message, original);
    }
}
