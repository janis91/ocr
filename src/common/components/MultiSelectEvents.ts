export class MultiSelectEvents {
    public static UPDATE_LANGS = 'update-langs';
    public static CLEAR_ALL = 'clear-all';
}
