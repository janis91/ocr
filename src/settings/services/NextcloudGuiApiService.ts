import { Translations } from '@/common/Translations'

export class NextcloudGuiApiService {
    /**
     * Displays an error for a given message.
     * @param message The message to display.
     */
    public static displayError: (message: string) => void = (message) => {
      OC.Notification.showHtml(`<div>${Translations.TRANSLATION_OCR}: ${message}</div>`, { timeout: 10, type: 'error' })
    }
}
