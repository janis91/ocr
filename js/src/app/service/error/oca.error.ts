import { OcrError } from './ocr.error';

export class OcaError extends OcrError {
    constructor(message: string, original?: Error | any) {
        super(message, original);
    }
}
