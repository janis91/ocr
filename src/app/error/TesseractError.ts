import { OcrError } from '@a/error/OcrError'

export class TesseractError extends OcrError {
}
