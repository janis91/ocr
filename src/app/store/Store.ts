import { generateUrl } from '@nextcloud/router'
import axios from '@nextcloud/axios'
import { AxiosResponse } from 'axios'
import { NextcloudGuiApiService } from '@a/services/NextcloudGuiApiService'
import { Translations } from '@a/configuration/Translations'
import { Util } from '@/common/Util'
import { StoreOptions } from 'vuex'

export const SHOW_MODAL = 'SHOW_MODAL'
export const HIDE_MODAL = 'HIDE_MODAL'
export const LOAD_FAVORITE_LANGUAGES = 'LOAD_FAVORITE_LANGUAGES'
const SET_FAVORITE_LANGUAGES = 'SET_FAVORITE_LANGUAGES'
export const UPDATE_SELECTED_LANGUAGES = 'UPDATE_SELECTED_LANGUAGES'
export const CLEAR_SELECTED_LANGUAGES = 'CLEAR_SELECTED_LANGUAGES'
export const SET_PROCESSING = 'SET_PROCESSING'
const SET_FINISHED = 'SET_FINISHED'
export const STOP_PROCESS_AND_RESET_MODAL_STATE = 'STOP_PROCESS_AND_RESET_MODAL_STATE'
export const SET_SELECTED_FILES = 'SET_SELECTED_FILES'
export const CLEAR_SELECTED_FILES = 'CLEAR_SELECTED_FILES'

export interface State {
  selectedFiles: OCAFile[];
  favoriteLanguages: Map<string, string>;
  showModal: boolean;
  selectedLanguages: Map<string, string>;
  processing: boolean;
}

export default {
  strict: true,
  state: {
    selectedFiles: [] as OCAFile[],
    favoriteLanguages: new Map<string, string>(),
    showModal: false as boolean,
    selectedLanguages: new Map<string, string>(),
    processing: false as boolean
  },
  getters: {
    selectedOptions(state) {
      return Util.mapToOptions(state.selectedLanguages)
    }
  },
  mutations: {
    [SHOW_MODAL](state) {
      state.showModal = true
    },
    [HIDE_MODAL](state) {
      state.showModal = false
    },
    [SET_PROCESSING](state) {
      state.processing = true
    },
    [SET_FINISHED](state) {
      state.processing = false
    },
    [SET_SELECTED_FILES](state, payload: {selectedFiles: OCAFile[]}) {
      state.selectedFiles = payload.selectedFiles
    },
    [CLEAR_SELECTED_FILES](state) {
      state.selectedFiles = []
    },
    [SET_FAVORITE_LANGUAGES](state, payload: { favoriteLanguages: Map<string, string> }) {
      state.favoriteLanguages = payload.favoriteLanguages
      state.selectedLanguages = payload.favoriteLanguages
    },
    [UPDATE_SELECTED_LANGUAGES](state, payload: { selectedLanguages: Map<string, string> }) {
      state.selectedLanguages = payload.selectedLanguages
    },
    [CLEAR_SELECTED_LANGUAGES](state) {
      state.selectedLanguages = new Map()
    }
  },
  actions: {
    async [LOAD_FAVORITE_LANGUAGES]({ commit }) {
      try {
        const { data } = await axios.get<any, AxiosResponse<string[]>>(generateUrl('/apps/ocr/api/personal/languages'))
        const favoriteLanguages = Util.mapToCommonLanguages(data)
        commit(SET_FAVORITE_LANGUAGES, { favoriteLanguages })
      } catch (e) {
        // eslint-disable-next-line no-console
        console.error('Unexpected error while loading favorite languages', e)
        NextcloudGuiApiService.displayError(Translations.TRANSLATION_UNEXPECTED_ERROR_LOAD_FAVORITE_LANGUAGES)
      }
    },
    [STOP_PROCESS_AND_RESET_MODAL_STATE]({ commit, state }) {
      commit(SET_FINISHED)
      commit(HIDE_MODAL)
      commit(UPDATE_SELECTED_LANGUAGES, { selectedLanguages: state.favoriteLanguages })
    }
  }
} as StoreOptions<State>
