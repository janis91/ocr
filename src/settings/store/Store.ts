import Vue from 'vue'
import Vuex, { Store } from 'vuex'
import { generateUrl } from '@nextcloud/router'
import axios from '@nextcloud/axios'
import { AxiosResponse } from 'axios'
import { NextcloudGuiApiService } from '@s/services/NextcloudGuiApiService'
import { Translations } from '@s/configuration/Translations'
import { Util, LanguageOption } from '@/common/Util'

Vue.use(Vuex)

export const SAVE_FAVORITE_LANGUAGES = 'SAVE_FAVORITE_LANGUAGES'
export const LOAD_FAVORITE_LANGUAGES = 'LOAD_FAVORITE_LANGUAGES'
export const SET_FAVORITE_LANGUAGES = 'SET_FAVORITE_LANGUAGES'
export const CLEAR_FAVORITE_LANGUAGES = 'CLEAR_FAVORITE_LANGUAGES'
const SET_STATUS_LOAD = 'SET_STATUS_LOAD'
const SET_STATUS_READY = 'SET_STATUS_READY'

export default new Store({
  strict: true,
  state: {
    favoriteLanguages: new Map<string, string>(),
    loading: true as boolean
  },
  mutations: {
    [SET_FAVORITE_LANGUAGES](state, payload: {favoriteLanguages: Map<string, string>}) {
      state.favoriteLanguages = payload.favoriteLanguages
    },
    [SET_STATUS_LOAD](state) {
      state.loading = true
    },
    [SET_STATUS_READY](state) {
      state.loading = false
    },
    [CLEAR_FAVORITE_LANGUAGES](state) {
      state.favoriteLanguages = new Map()
    }
  },
  getters: {
    selectedOptions(state): LanguageOption[] {
      return Util.mapToOptions(state.favoriteLanguages)
    }
  },
  actions: {
    async [LOAD_FAVORITE_LANGUAGES]({ commit }) {
      try {
        commit(SET_STATUS_LOAD)
        const { data } = await axios.get<any, AxiosResponse<string[]>>(generateUrl('/apps/ocr/api/personal/languages'))
        const favoriteLanguages = Util.mapToCommonLanguages(data)
        commit(SET_FAVORITE_LANGUAGES, { favoriteLanguages })
        commit(SET_STATUS_READY)
      } catch (e) {
        NextcloudGuiApiService.displayError(Translations.TRANSLATION_UNEXPECTED_ERROR_LOAD)
        commit(SET_FAVORITE_LANGUAGES, { favoriteLanguages: new Map() })
        commit(SET_STATUS_READY)
      }
    },
    async [SAVE_FAVORITE_LANGUAGES]({ commit, dispatch, state }) {
      try {
        commit(SET_STATUS_LOAD)
        const dataToTransfer = { favoriteLanguages: JSON.stringify(Array.from(state.favoriteLanguages.keys())) }
        const { data } = await axios.post<{ favoriteLanguages: string }, AxiosResponse<string[]>>(generateUrl('/apps/ocr/api/personal/languages'), dataToTransfer)
        const favoriteLanguages = Util.mapToCommonLanguages(data)
        commit(SET_FAVORITE_LANGUAGES, { favoriteLanguages })
        commit(SET_STATUS_READY)
      } catch (e) {
        if (e.response.status === 400) {
          NextcloudGuiApiService.displayError(Translations.TRANSLATION_ERROR_WRONG_INPUT)
        } else {
          NextcloudGuiApiService.displayError(Translations.TRANSLATION_UNEXPECTED_ERROR_SAVE)
        }
        await dispatch(LOAD_FAVORITE_LANGUAGES)
      }
    }
  }
})
