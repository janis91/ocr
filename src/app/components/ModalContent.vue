<template>
  <multi-select
    :disabled="disabled"
    :selectedLanguages="selectedLanguages"
    @[update]="updateSelectedLanguages"
    @[clear]="clearAll"
  ></multi-select>
</template>

<script lang="ts">
import Vue from 'vue'
import MultiSelect, { MultiSelectEvents } from '@/common/components/MultiSelect.vue'
import { LanguageOption } from '@/common/Util'
import { UPDATE_SELECTED_LANGUAGES, CLEAR_SELECTED_LANGUAGES } from '@a/store/Store'

export default Vue.extend({
  name: 'ModalContent',
  components: { MultiSelect },
  data: () => ({
    update: MultiSelectEvents.UPDATE_LANGS,
    clear: MultiSelectEvents.CLEAR_ALL
  }),
  computed: {
    selectedFiles(): OCAFile[] {
      return this.$store.state.selectedFiles
    },
    disabled(): boolean {
      return this.$store.state.processing
    },
    selectedLanguages(): LanguageOption[] {
      return this.$store.getters.selectedOptions
    }
  },
  methods: {
    updateSelectedLanguages(value: LanguageOption[]) {
      this.$store.commit(UPDATE_SELECTED_LANGUAGES, { selectedLanguages: value })
    },
    clearAll() {
      this.$store.commit(CLEAR_SELECTED_LANGUAGES)
    }
  }
})
</script>

<style scoped>
</style>
