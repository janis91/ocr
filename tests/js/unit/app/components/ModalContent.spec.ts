import '../../GlobalMocks.mock'
import { shallowMount, Wrapper, createLocalVue } from '@vue/test-utils'
import Vue from 'vue'
import Vuex from 'vuex'
import ModalContent from '@a/components/ModalContent.vue'
import { MultiSelectEvents } from '@/common/components/MultiSelectEvents'
import { FileFixtures } from '../../fixtures/FileFixtures'
import { STOP_PROCESS_AND_RESET_MODAL_STATE, State, UPDATE_SELECTED_LANGUAGES, CLEAR_SELECTED_LANGUAGES, SET_PROCESSING } from '@a/store/Store'
import { NextcloudGuiApiService } from '@a/services/NextcloudGuiApiService'
import { TesseractError } from '@a/error/TesseractError'

jest.mock('@a/services/NextcloudGuiApiService')

describe('ModalContent.vue', () => {
  let wrapper: Wrapper<any>
  let localVue: typeof Vue
  let actions: any
  let getters: any
  let mutations: any
  let state: Partial<State>

  beforeEach(() => {
    localVue = createLocalVue()
    localVue.use(Vuex)
    state = {
      processing: false,
      selectedFiles: [FileFixtures.PNG]
    }
    mutations = {
      [UPDATE_SELECTED_LANGUAGES]: jest.fn(),
      [CLEAR_SELECTED_LANGUAGES]: jest.fn(),
      [SET_PROCESSING]: jest.fn()
    }
    actions = {
      [STOP_PROCESS_AND_RESET_MODAL_STATE]: jest.fn()
    }
    getters = {
      selectedOptions: () => [{ key: 'deu', label: 'German' }]
    }
    const store = new Vuex.Store({
      strict: true,
      state,
      actions,
      getters,
      mutations
    })
    wrapper = shallowMount(ModalContent, { store, localVue });
    (OCA.Ocr as any) = {
      app: {
        process: jest.fn(() => { throw new Error('Mock behavior not defined') })
      }
    }
  })

  it('GIVEN standard state, WHEN ModalContent is mounted, THEN matches snapshot with button to trigger process and without loading indicator.', async () => {
    expect(wrapper.contains('button')).toBeTruthy()
    expect(wrapper.contains('loading')).toBeFalsy()
    expect(wrapper.find('check-stub').attributes()['disabled']).toBeFalsy()
    expect(wrapper.find('multi-select-stub').attributes()['disabled']).toBeFalsy()
    expect(wrapper.element).toMatchSnapshot()
  })

  it('GIVEN state with processing = true, WHEN ModalContent is mounted, THEN matches snapshot with loading indicator instead of button.', async () => {
    state.processing = true
    const store = new Vuex.Store({
      strict: true,
      state,
      actions,
      getters,
      mutations
    })
    wrapper = shallowMount(ModalContent, { store, localVue })
    expect(wrapper.contains('button')).toBeFalsy()
    expect(wrapper.contains('loading-stub')).toBeTruthy()
    expect(wrapper.find('check-stub').attributes()['disabled']).toBeTruthy()
    expect(wrapper.find('multi-select-stub').attributes()['disabled']).toBeTruthy()
    expect(wrapper.element).toMatchSnapshot()
  })

  it('GIVEN standard state, WHEN ModalContent is mounted AND check-stub @input emits true, THEN replace is set to true.', async () => {
    expect(wrapper.vm.replace).toBeFalsy()
    wrapper.find('check-stub').vm.$emit('input', true)

    await wrapper.vm.$nextTick()

    expect(wrapper.vm.replace).toBeTruthy()
  })

  it('GIVEN standard state, WHEN ModalContent is mounted AND multi-select @update-langs emits Map(), THEN correpsonding store mutation is commited.', async () => {
    wrapper.find('multi-select-stub').vm.$emit(MultiSelectEvents.UPDATE_LANGS, new Map())

    await wrapper.vm.$nextTick()

    expect(mutations[UPDATE_SELECTED_LANGUAGES]).toHaveBeenCalledWith(state, { selectedLanguages: new Map() })
  })

  it('GIVEN standard state, WHEN ModalContent is mounted AND multi-select @clear-all emits, THEN correpsonding store mutation is commited.', async () => {
    wrapper.find('multi-select-stub').vm.$emit(MultiSelectEvents.CLEAR_ALL)

    await wrapper.vm.$nextTick()

    expect(mutations[CLEAR_SELECTED_LANGUAGES]).toHaveBeenCalledWith(state, undefined)
  })

  it('GIVEN standard state AND ocr process resolves, WHEN ModalContent is mounted AND button is clicked, THEN corresponding actions and process is triggered.', async () => {
    (OCA.Ocr.app.process as jest.Mock).mockImplementationOnce(() => Promise.resolve())
    wrapper.vm.replace = true
    expect(wrapper.vm.successfullyProcessed).toBe(0)

    wrapper.find('button').trigger('click')

    await wrapper.vm.$nextTick()
    await wrapper.vm.$nextTick()

    expect(mutations[SET_PROCESSING]).toHaveBeenCalled()
    expect(actions[STOP_PROCESS_AND_RESET_MODAL_STATE]).toHaveBeenCalled()
    expect(wrapper.vm.successfullyProcessed).toBe(0)
    expect(wrapper.vm.replace).toBeFalsy()
    expect(OCA.Ocr.app.process).toHaveBeenCalledWith(FileFixtures.PNG, ['deu'], true)
  })

  it('GIVEN standard state AND ocr process rejects, WHEN ModalContent is mounted AND button is clicked, THEN corresponding actions are triggered and displays error.', async () => {
    (OCA.Ocr.app.process as jest.Mock).mockImplementationOnce(() => Promise.reject(new TesseractError('test')));
    (NextcloudGuiApiService.displayError as jest.Mock).mockImplementationOnce(() => {})
    wrapper.vm.replace = false
    wrapper.vm.successfullyProcessed = 1

    wrapper.find('button').trigger('click')

    await wrapper.vm.$nextTick()
    await wrapper.vm.$nextTick()

    expect(mutations[SET_PROCESSING]).toHaveBeenCalled()
    expect(actions[STOP_PROCESS_AND_RESET_MODAL_STATE]).toHaveBeenCalled()
    expect(wrapper.vm.successfullyProcessed).toBe(0)
    expect(wrapper.vm.replace).toBeFalsy()
    expect(OCA.Ocr.app.process).toHaveBeenCalledWith(FileFixtures.PNG, ['deu'], false)
    expect(NextcloudGuiApiService.displayError).toHaveBeenCalledWith('OCR processing failed: test')
  })
})
