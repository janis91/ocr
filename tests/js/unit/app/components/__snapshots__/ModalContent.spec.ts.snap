// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalContent.vue GIVEN standard state, WHEN ModalContent is mounted, THEN matches snapshot with button to trigger process and without loading indicator. 1`] = `
<div
  class="ocr-content"
>
  <file-list-stub
    files="[object Object]"
  />
   
  <div
    class="ocr-options"
  >
    <multi-select-stub
      selectedlanguages="[object Object]"
    />
     
    <check-stub
      filescount="1"
    />
     
    <hint-stub
      hint="A large number of files may take a very long time."
    />
  </div>
   
  <!---->
   
  <button>
    Process
  </button>
</div>
`;

exports[`ModalContent.vue GIVEN state with processing = true, WHEN ModalContent is mounted, THEN matches snapshot with loading indicator instead of button. 1`] = `
<div
  class="ocr-content"
>
  <file-list-stub
    files="[object Object]"
  />
   
  <div
    class="ocr-options"
  >
    <multi-select-stub
      disabled="true"
      selectedlanguages="[object Object]"
    />
     
    <check-stub
      disabled="true"
      filescount="1"
    />
     
    <hint-stub
      hint="A large number of files may take a very long time."
    />
  </div>
   
  <loading-stub
    filesqueued="1"
    successfullyprocessed="0"
  />
   
  <!---->
</div>
`;
